File: queues/redis/Ruby/lib/redisclient.rb
# Copyright Core DF
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'json'
require_relative 'result'

module Redisclient
  module_function

  def connection_url
    return ENV['REDIS_URL'] if ENV['REDIS_URL'] && !ENV['REDIS_URL'].empty?
    host = ENV['REDIS_HOST'].to_s
    return '' if host.empty?
    port = ENV.fetch('REDIS_PORT', '6379')
    db = ENV.fetch('REDIS_DB', '0')
    pass = ENV['REDIS_PASSWORD'].to_s
    pass.empty? ? "redis://\#{host}:\#{port}/\#{db}" : "redis://:\#{pass}@\#{host}:\#{port}/\#{db}"
  end

  def encode(value)
    return value if value.is_a?(String)
    return JSON.generate(value) if value.is_a?(Hash) || value.is_a?(Array)
    value.to_s
  end

  def decode(raw)
    JSON.parse(raw)
  rescue StandardError
    raw.to_s
  end

  def Init
    return CoreautoResult.missing_env('REDIS_URL or REDIS_HOST') if connection_url.empty?
    { status_code: 200 }
  end

  def Push(queue, value)
    return CoreautoResult.missing_env('REDIS_URL or REDIS_HOST') if connection_url.empty?
    require 'redis'
    Redis.new(url: connection_url).lpush(queue, encode(value))
    { status_code: 200 }
  rescue LoadError
    { status_code: 500, error: 'redis gem required' }
  rescue StandardError => e
    CoreautoResult.transport_error(e.message)
  end

  def Pop(queue, timeout_sec: 30, max_messages: 1)
    return CoreautoResult.missing_env('REDIS_URL or REDIS_HOST') if connection_url.empty?
    require 'redis'
    r = Redis.new(url: connection_url)
    messages = []
    remaining = [max_messages, 1].max
    while remaining > 0
      wait = remaining == max_messages ? [timeout_sec.to_i, 1].max : 1
      item = r.brpop(queue, timeout: wait)
      break unless item
      _key, body = item
      messages << { queue: queue, value: decode(body) }
      remaining -= 1
      timeout_sec -= wait
      break if timeout_sec <= 0
    end
    { status_code: 200, messages: messages }
  rescue LoadError
    { status_code: 500, error: 'redis gem required' }
  rescue StandardError => e
    CoreautoResult.transport_error(e.message)
  end
end

File: queues/redis/Rust/src/redisclient.rs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::result::{missing_env, transport_error};
use redis::Commands;
use serde_json::{json, Value};
use std::env;

fn connection_url() -> String {
    if let Ok(url) = env::var("REDIS_URL") {
        if !url.is_empty() {
            return url;
        }
    }
    let host = env::var("REDIS_HOST").unwrap_or_default();
    if host.is_empty() {
        return String::new();
    }
    let port = env::var("REDIS_PORT").unwrap_or_else(|_| "6379".into());
    let db = env::var("REDIS_DB").unwrap_or_else(|_| "0".into());
    let pass = env::var("REDIS_PASSWORD").unwrap_or_default();
    if pass.is_empty() {
        format!("redis://{host}:{port}/{db}")
    } else {
        format!(
            "redis://:{}@{}:{}/{}",
            urlencoding::encode(&pass),
            host,
            port,
            db
        )
    }
}

fn encode(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn decode(raw: &str) -> Value {
    serde_json::from_str(raw).unwrap_or(Value::String(raw.to_string()))
}

pub fn init() -> Value {
    if connection_url().is_empty() {
        return missing_env("REDIS_URL or REDIS_HOST");
    }
    json!({"status_code":200})
}

pub fn push(queue: &str, value: Value) -> Value {
    let url = connection_url();
    if url.is_empty() {
        return missing_env("REDIS_URL or REDIS_HOST");
    }
    let payload = encode(&value);
    let client = match redis::Client::open(url.as_str()) {
        Ok(c) => c,
        Err(e) => return transport_error(&e.to_string()),
    };
    let mut con = match client.get_connection() {
        Ok(c) => c,
        Err(e) => return transport_error(&e.to_string()),
    };
    match con.lpush::<_, _, ()>(queue, payload) {
        Ok(()) => json!({"status_code":200}),
        Err(e) => transport_error(&e.to_string()),
    }
}

pub fn pop(queue: &str, timeout_sec: f64, max_messages: usize) -> Value {
    let url = connection_url();
    if url.is_empty() {
        return missing_env("REDIS_URL or REDIS_HOST");
    }
    let client = match redis::Client::open(url.as_str()) {
        Ok(c) => c,
        Err(e) => return transport_error(&e.to_string()),
    };
    let mut con = match client.get_connection() {
        Ok(c) => c,
        Err(e) => return transport_error(&e.to_string()),
    };
    let mut messages = Vec::new();
    let mut remaining = max_messages.max(1);
    let mut deadline = timeout_sec;
    while remaining > 0 {
        let wait = if remaining == max_messages {
            deadline.max(1.0) as f64
        } else {
            1.0
        };
        match con.brpop::<_, Option<(String, String)>>(queue, wait) {
            Ok(Some((_key, body))) => {
                messages.push(json!({"queue": queue, "value": decode(&body)}));
                remaining -= 1;
                deadline -= wait;
                if deadline <= 0.0 {
                    break;
                }
            }
            Ok(None) => break,
            Err(e) => return transport_error(&e.to_string()),
        }
    }
    json!({"status_code":200,"messages":messages})
}
